### ISSUE TITLE
use : acquire_token_by_username_password

### ISSUE THREAD

**Issue by hitokiri** [O365/python-o365#801]
Hello, for security reasons, I need to use the user and passsword to generate the token, I have done it with the MSAL library, but I wanted to know if there are plans to implement this in this library?

`token = self.app.acquire_token_by_username_password(username=self.username,
                                                    password=self.password,
                                                    scopes=['.default'])`

**Comment by janscas** [comment]
We don't use MSAL although I'm very aware that we should use it, but I don't have the time to implement this.

We currently don't have any option to adquire the token by username and password.

**Comment by hitokiri** [comment]
Hi @janscas , I know you don't have time, and I thank you for your work with this library, I just came up with that, I made a fix there by creating the token with the other library, and it worked perfectly, with yours, I had doubts if you were planning to aimplement that, that's why the question, but thanks again for your work.

**Comment by abubelinha** [comment]
Sorry but I am a newcomer and I really get lost on how to access my account with this library.
I have a personal email account at my university (which uses Office 365 services) and I just know my username and password.
I would just like to read all my emails using Python and regex-filter subjects to redirect some of them to a different account (gmail).

I looked into https://o365.github.io/python-o365/latest/usage/account.html#authenticating-your-account
But I am so confused. I don't understand where do those `'my_client_id', 'my_client_secret'` come from.

For what @janscas says I understand I cannot use this O365 library to access my emails from Python if I just know my username and password?
But @hitokiri says I can't if I combining O365 with "the other library"?

Could any of you provide details about what that "other library" means, please?
Or any explanation on how to get `'my_client_id', 'my_client_secret'` starting from `'my_username', 'my_password'` (in case that's possible at all).

Thanks a lot in advance
@abubelinha

**Comment by janscas** [comment]
Follow the readme it’s very well explained. The client secret and id comes from the app set up in azure portal. Check the readme and try

**Comment by abubelinha** [comment]
Thanks again for your quick answer @janscas 
With "the app set up in azure portal" you mean these [_authentication steps_](https://github.com/O365/python-o365#authentication-steps)?

> To allow authentication you first need to register your application at [Azure App Registrations](https://portal.azure.com/#blade/Microsoft_AAD_RegisteredApps/ApplicationsListBlade).
> 
> Login at [Azure Portal (App Registrations)](https://portal.azure.com/#blade/Microsoft_AAD_RegisteredApps/ApplicationsListBlade)
> 
> Create an app. Set a name.

So, in that "Azure login" is where I should use my university account username & address?
I get this screen:

> You do not have access
> 
> Your administrator has disabled the App registrations experience in the Azure portal. You can still register or manage applications using PowerShell or another client such as Visual Studio.
> 

So, this means end of O365 story for me?  :disappointed: 
What's that "PowerShell or Visual Studio" alternative proposed?

**Comment by janscas** [comment]
Yep … ask the admin if they can provide access to set up an app or if they can set up one for you. That’s the only way

**Comment by abubelinha** [comment]
No way. Not very friendly staff, and they don't even answer phone or emails.

It's a shame. I just want to redirect messages to my Gmail, which suddenly stopped to grab them from other POP3 accounts (at least from outlook accounts).  Do you know of any other Python libraries to just read an outlook.office365.com POP account?

**Comment by janscas** [comment]
Not that I know…

**Comment by abubelinha** [comment]
Thanks @jancas. 
I know this is not related to the original issue, but I'd have a more general question to understand how the authentication works.

Figure out what I am trying to do is a Python CLI application that anyone can use for reading his/her email (not just for me). 
I mean, something like Eudora or Mozilla Thunderbird desktop applications, that anybody can download, install, and configure pop/smtp servers plus his/her account username/password.

My University lets us use those applications. 
But in those cases, I think this step of "app set up in azure portal" is not necessary.
I don't understand **how it is possible for these apps to authenticate my user, but impossible for O365 Python apps**.

Perhaps I can use a different Microsoft account to create the app (not my uni account), and then users authenticate in their own accounts through the app itself? (sorry I am talking about something I don't really understand how it works).

**Comment by janscas** [comment]
The method used by this library is by setting that “app”. If there’s another way and what limitations may have I don’t know. I can’t tell right now other means of reading email, modifying excel, calendar stuff, one drive and etc.

anyway if you just need email maybe pop3 or others I don’t know about

**Comment by abubelinha** [comment]
@hitokiri 
Please, would you be so kind to share the python code you employed to use O365 library by login using the MSAL and then sharing the token?

That would solve the issue for other people like me too.

Thanks a lot in advance!

**Comment by alejcas** [comment]
O365 2.1 now uses msal under the hood
